### PR TITLE
Raincatch-399 - User details passed back from the server should not include any password information.

### PR DESCRIPTION
**Pull request by tomjackman**
Currently the authentication response will return all fields on a user, which can include their password and other sensitive information. This has been changed to not return the password in the authentication response. Furthermore you can configure which fields should not be returned by specifying them in params.exclude, an array. Documentation was added to the readme module to cover these options and usage.

### CODE REVIEW

**Comment by tomjackman** [comment]
@nialldonnellyfh Can you have a quick look at this WIP PR? Still requires tests.

File: lib/router/mbaas.js [hunk 1]
```diff
@@ -86,6 +89,28 @@ function initRouter(mediator) {
...
+/**
+* Function to trim the authentication response to remove certain fields from being sent.
+* By default, the password will be removed from the response.
+* @param authResponse {object} - the untrimmed auth response
+* @param exclusionList {array} - the array of field names to remove from the authentication response
+* @return authResponse {object} - the trimmed authentication response
+*/
+function trimAuthResponse(authResponse, exclusionList) {
+  // remove the password field by default from the authentication response if no list of fields to exclude is specified
+  var excludedFields = exclusionList || ["password"];
```
Comment (nialldonnellyfh) [inline lib/router/mbaas.js]: The default can be part of the config for the module.

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "fh-wfm-user",
-  "version": "0.0.25",
+  "version": "0.0.26",
```
Comment (nialldonnellyfh) [inline package.json]: I'd say leave version bumping for when the module is being published as part of the release process.

**Comment by nialldonnellyfh** [comment]
@TommyJ1994 couple of comments above. Looks like it's shaping up nicely!

**Comment by tomjackman** [comment]
Moved options to config file. Updated Docs. Added Tests.

File: lib/router/mbaas.js [hunk 3]
```diff
@@ -86,7 +89,27 @@ function initRouter(mediator) {
...
 }
 
+/**
+* Function to trim the authentication response to remove certain fields from being sent.
+* By default, the password will be removed from the response.
+* @param authResponse {object} - the untrimmed auth response
+* @param exclusionList {array} - the array of field names to remove from the authentication response
+* @return authResponse {object} - the trimmed authentication response
+*/
+function trimAuthResponse(authResponse, exclusionList) {
```
Comment (ghost) [inline lib/router/mbaas.js:97]: you need to clone profileData(authResponse) first and then remove all excluded elements. better to have profileData immutable

File: lib/router/mbaas-spec.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  describe('#testAuthResponseData', function() {
+    it('it should not remove any fields when no exclusion list is specified', function(done) {
+      var authResponse = mbaas.trimAuthResponse(sampleProfileData, sampleExclusionList3);
+      assert(Object.keys(authResponse).length === sampleProfileDataLength);
+      done();
+    });
+    it('it should remove the password field by default', function(done) {
+      var authResponse = mbaas.trimAuthResponse(sampleProfileData, config.authResponseExclusionList);
+      assert(authResponse.password === undefined);
+      assert(Object.keys(authResponse).length !== sampleProfileDataLength);
```
Comment (ghost) [inline lib/router/mbaas-spec.js]: What returns assert(Object.keys(authResponse).length !== sampleProfileData.length);? It should be same

**Comment by tomjackman** [comment]
Ping @nialldonnellyfh can you take a look?

File: Gruntfile.js [hunk 5]
```diff
@@ -1,10 +1,24 @@
...
+
 module.exports = function(grunt) {
-  'use strict';
+  require('load-grunt-tasks')(grunt);
   grunt.initConfig({
+    pkg: grunt.file.readJSON('package.json'),
     eslint: {
       src: ["lib/**/*.js"]
+    },
+    shell: {
```
Comment (nialldonnellyfh) [inline Gruntfile.js]: Might want to use [grunt-mocha](https://github.com/kmiyashiro/grunt-mocha) instead of shelling out.

File: README.md [hunk 6]
```diff
@@ -116,10 +116,13 @@ Base url : `/api/wfm/user`
 
 | resource | method | returns |
 | -------- | ------ | ------- |
-| /auth | all | `{satus: 'ok', userId: username, sessionToken: sessiontoken, authResponse: profileData}` |
+| /auth | all | `{status: 'ok', userId: username, sessionToken: sessiontoken, authResponse: authResponse}` |
```
Comment (nialldonnellyfh) [inline README.md]: It would be good to explain what `sessiontoken` and `authResponse` are.

File: lib/router/mbaas-spec.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
+var _ = require('lodash');
+const assert = require('assert');
+const authHandler = require('./mbaas');
+const sampleUserConfig = require('../../test/fixtures/sampleUserConfig.json');
+const sampleProfileData = require('../../test/fixtures/sampleUserProfile.json');
+const sampleProfileDataLength  = Object.keys(sampleProfileData).length;
+
+var sampleExclusionList1 = ['banner'];
+var sampleExclusionList2 = ['banner', 'avatar'];
+var sampleExclusionList3 = [];
```
Comment (nialldonnellyfh) [inline lib/router/mbaas-spec.js:10]: What happens if `authResponseExclusionList` is `null` or `undefined` ? Would be good to test that.

File: lib/router/mbaas.js [hunk 8]
```diff
@@ -86,7 +89,27 @@ function initRouter(mediator) {
...
+  _.forEach(exclusionList, function(value) {
```
Comment (nialldonnellyfh) [inline lib/router/mbaas.js]: Might want to use https://lodash.com/docs/4.17.1#omit instead.

`delete` removes optimisation options. https://www.smashingmagazine.com/2012/11/writing-fast-memory-efficient-javascript/#de-referencing-misconceptions

File: lib/user/config-user.js [hunk 9]
```diff
@@ -4,5 +4,6 @@ module.exports = {
   apiHost: 'http://localhost:8080',
   apiPath: '/api/wfm/user',
   authpolicyPath: '/box/srv/1.1/admin/authpolicy',
-  policyId: process.env.WFM_AUTH_POLICY_ID || 'wfm'
+  policyId: process.env.WFM_AUTH_POLICY_ID || 'wfm',
+  authResponseExclusionList: ['password']
```
Comment (nialldonnellyfh) [inline lib/user/config-user.js]: Can this be configured externally by the consuming application?

**Review by nialldonnellyfh — COMMENTED** [review]
Couple of comments.

It would be good to have the options be configurable by the application also.

Otherwise, every time a config value has to change, a new version of the module would have to be used.

**Review by vchepeli — COMMENTED** [review]
LGTM, small fixes

File: lib/router/mbaas-spec.js (hunk 4 above)
Comment (tomjackman) [reply lib/router/mbaas-spec.js]: By default the password field is removed if no exclusion list is provided. Therefore the original profile data should have a different length to the authResponse. (ie. password removed from authResponse)

File: README.md [hunk 10]
```diff
@@ -114,12 +114,15 @@ Base url : `/api/wfm/[group|user|membership|`
...
-| /auth | all | `{satus: 'ok', userId: username, sessionToken: sessiontoken, authResponse: profileData}` |
-| /verifysession | all | `{isValid: true}` |
-| /revokesession | all | `{}` |
+| /auth | all | `{status: 'ok', userId: username, sessionToken: sessiontoken, authResponse: authResponse}` | `sessionToken` : identifier for a specific user for the duration of that user's visit. <br>  `authResponse` : authentication response containing the authenticated users details. |
+| /verifysession | all | `{isValid: true}` | |
+| /revokesession | all | `{}` |  | |
+
+*Note*: You can choose which fields to remove from the authentication response by specifying the field names in the `authResponseExclusionList` in `lib/user/config-user.js`.
 
+For Example, setting the array as `['password', 'banner']` will prevent these fields from appearing in the authentication response. By default, the `password` field is removed from the response. To allow all fields to be sent, set `authResponseExclusionList` as an empty array.
```
Comment (nialldonnellyfh) [inline README.md]: You also need to update the "Setup mbaas service" section for the exclusion list. This text section should be there.

File: lib/router/mbaas.js (hunk 3 above)
Comment (nialldonnellyfh) [reply lib/router/mbaas.js:97]: @vchepeli It's not changed. https://lodash.com/docs/4.17.2#omit will return  a different object.

**Review by nialldonnellyfh — COMMENTED** [review]
Looks good. Just needs testing in a cluster.

**Comment by paolobueno** [comment]
Closing in favor of #25

**Pull request closed without merging**